### PR TITLE
Spin origin

### PR DESCRIPTION
**Pull request by julient31**
## Purpose
Pull request of the implementation of the SPIN package. Enabling the simulation of coupled spin dynamics and molecular dynamics. 

## Author(s)
Julien Tranchida (Sandia), Aidan P. Thompson (Sandia)

## Implementation Notes
- the changes were mainly implemented within the src/SPIN repository
- the implementation was checked by series of Valgrind tests (running on all the examples)
- different compilation options were tested (-DLAMMPS_SMALLSMALL and -DLAMMPS_BIGBIG, -O -Wall -Wextra) 
- the implementation was compiled with different compilers (gcc, intel C++ and Clang++)

## Further Information, Files, and Links
Details of the implementations are provided within the following article: 
https://arxiv.org/abs/1801.10233

### CODE REVIEW

File: src/dump_custom.cpp [hunk 1]
```diff
@@ -200,8 +201,8 @@ DumpCustom::~DumpCustom()
 
   delete [] pack_choice;
   delete [] vtype;
-  memory->destroy(field2index);
-  memory->destroy(argindex);
+  delete [] field2index;
```
Comment (stanmoore1) [inline src/dump_custom.cpp]: Needs to switch back to use memory class.

File: src/atom.cpp [hunk 2]
```diff
@@ -492,7 +507,7 @@ AtomVec *Atom::new_avec(const char *style, int trysuffix, int &sflag)
     AtomVecCreator avec_creator = (*avec_map)[style];
     return avec_creator(lmp);
   }
-
+  //printf("test entries function: %s, %d, %d \n ",style, trysuffix, &sflag);
```
Comment (stanmoore1) [inline src/atom.cpp]: Remove debug code.

File: src/dump_custom.cpp [hunk 3]
```diff
@@ -244,15 +245,11 @@ DumpCustom::~DumpCustom()
   for (int i = 1; i <= ntypes; i++) delete [] typenames[i];
   delete [] typenames;
 
-  if(vformat) {
-    for (int i = 0; i < size_one; i++) delete [] vformat[i];
-    delete [] vformat;
-  }
+  for (int i = 0; i < size_one; i++) delete [] vformat[i];
```
Comment (stanmoore1) [inline src/dump_custom.cpp]: Add back in the if statement, or justify why it was removed.

File: src/verlet.cpp [hunk 4]
```diff
@@ -385,6 +387,8 @@ void Verlet::force_clear()
 
     if (nbytes) {
       memset(&atom->f[0][0],0,3*nbytes);
+      //test memset for fm
+      //memset(&atom->fm[0][0],0,3*nbytes);
```
Comment (stanmoore1) [inline src/verlet.cpp]: Remove debug code.

**Comment by akohlmey** [comment]
@julient31 am working through a bunch of issues most likely related to your rebasing onto a newer LAMMPS master. in the process the file `doc/src/tutorial_spin.txt seems to have gotten lost. i'll remove the reference to it now. you can add it in a future pull request after this one is merged properly into master.

**Pull request merged**